### PR TITLE
Pass render options and block to calls to `#render_in`

### PR DESCRIPTION
**Pull request by seanpdoyle**
### Motivation / Background

Closes [#45432][]

Support for objects that respond to `#render_in` was introduced in [#36388][] and [#37919][]. Those implementations assume that the instance will all the context it needs to render itself. That assumption doesn't account for call-site arguments like `locals: { ... }` or a block.

### Detail

This commit expands support for rendering with a `:renderable` option to incorporate locals and blocks. For example:

```ruby
class Greeting
  def render_in(view_context, **options, &block)
    if block
      view_context.render html: block.call
    else
      view_context.render inline: <<~ERB.strip, **options
        Hello, <%= name %>
      ERB
    end
  end
end

render(Greeting.new)                    # => "Hello, World"
render(Greeting.new, name: "Local")     # => "Hello, Local"
render(Greeting.new) { "Hello, Block" } # => "Hello, Block"
```

Since existing tools depend on the `#render_in(view_context)` signature
without options, this commit deprecates that signature in favor of one
that accepts options and a block.

[#45432]: https://github.com/rails/rails/pull/45432
[#36388]: https://github.com/rails/rails/pull/36388
[#37919]: https://github.com/rails/rails/pull/37919

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by seanpdoyle** [comment]
The support for blocks was somewhat incidental here. My main goal was to support passing `:locals` provided at the call sites.

@joelhawksley (and @BlakeWilliams, since you've authored https://github.com/rails/rails/pull/45432) does this change support behavior that's desirable to ViewComponents?

**Comment by BlakeWilliams** [comment]
Thanks for adding block support! I lost that one in my TODO's a while back since I've been pretty busy lately.

I'm not positive the locals changes as-is would help us much since we expect all data to be passed in to components via the constructor, but also with the new API it's no longer clear to me when to pass something as a local vs constructor argument. e.g.

```ruby
render(Greeting.new(name: "Fox Mulder"), name: "Dana Scully"))
```

I think the API changes make that valid code, but it's not clear to me which argument would take precedence or how Greeting should respond to locals from a ViewComponent perspective (we use `render_in` as a hook into ActionView, but otherwise components are largely just Ruby including the constructor). Coming at this with a relatively heavy ViewComponent bias, I could see an API that delegates to the constructor making sense, maybe like:

```ruby
render(Greeting.new(name: "local"))
render(Greeting, name: "local") # Equivalent to the above since Greeting has a `render_in` instance_method. `name:` is passed to the constructor
```

All of that has a heavy ViewComponent bias of course, but I'm curious if you had any particular use-cases for the locals changes in mind, ViewComponent or otherwise?

**Comment by seanpdoyle** [comment]
> I'm curious if you had any particular use-cases for the locals changes in mind, ViewComponent or otherwise?

At the moment, the main motivating use case is internal to Rails. I'm in the process of exploring an Action Text branch that adds support for editors other than Trix.

Incidentally, one part of the code I'm exploring re-structuring involves the [ActionText::ContentHelper](https://github.com/rails/rails/blob/b294a25a8a59abef00a72700607b53bac901f83d/actiontext/app/helpers/action_text/content_helper.rb) helper module, namely the [`render_action_text_attachment` method:

https://github.com/rails/rails/blob/b294a25a8a59abef00a72700607b53bac901f83d/actiontext/app/helpers/action_text/content_helper.rb#L43-L55

It's a `:nodoc:` method, and is only called within the module itself:

https://github.com/rails/rails/blob/b294a25a8a59abef00a72700607b53bac901f83d/actiontext/app/helpers/action_text/content_helper.rb#L30

https://github.com/rails/rails/blob/b294a25a8a59abef00a72700607b53bac901f83d/actiontext/app/helpers/action_text/content_helper.rb#L36

The implementation is fairly concerned with checking whether or not the `attachment` responds to various methods, all for the sake of building up the correct `Hash` or options to pass to `render`. One extra layer on top of that is that between the two call sites, the only difference is the `:in_gallery` local variable (`true` in one place, `false` in the other).

I'm imagining a world where each editor adapter (ckeditor5, tinymce, etc) have an opportunity to subclass `ActionText::Attachment` and want to override the default behavior. They wouldn't be able (or advised to) reach into the module-private `:nodoc:` method. However, if `render_in` accepted options, they'd be available to the subclasses at call time.

That's a circumstance where the constructor arguments to the object and the values passed into the `:locals` wouldn't be similar enough to cause confusion. In this use case, it's more about the context with which they're rendered. The `ActionText::Attachment` objects are constructed in far-flung parts of Action Text very separate from the HTML rendering `render_in` calls. It would also provide the framework with an opportunity to maintain the consistency in the construction of the `ActionText::Attachment` instances while providing integrators with flexibility at render-time.

Outside of that use case, I'm imagining other scenarios where an application might want to build renderable objects (maybe even ViewComponents) prior to their render. I don't have a concrete example handy, but I'm imagining a component that yields itself into another component without knowledge of how the receiving component would want to treat it.

**Comment by seanpdoyle** [comment]
@BlakeWilliams if this change is accepted, I'm also exploring adding a default `#render_in` implementation to the `ActiveModel::Conversion` module:

https://github.com/seanpdoyle/rails/compare/action-view-render-in-options...seanpdoyle:rails:active-model-conversion-render-in#diff-8ec888bfb07b8d4a67ffa913e3286d4ffcb052d1bf3b232d2874c7177cd96fbc

The idea there would be that any Active Model or Active Record instance would provide a seam for applications to take over control of rendering (for example, mapping a `Person` to a `PersonComponent` class).

File: actionview/lib/action_view/template/renderable.rb [hunk 1]
```diff
@@ -4,20 +4,22 @@ module ActionView
...
       end
 
       def identifier
         @renderable.class.name
       end
 
-      def render(context, *args)
-        @renderable.render_in(context)
+      def render(context, locals)
+        @renderable.render_in(context, formats: @formats, locals: locals, &@block)
```
Comment (rafaelfranca) [inline actionview/lib/action_view/template/renderable.rb:18]: This is a breaking change for those objets. We probably need to deprecate the old signature and tell people to upgrade their objects

File: actionview/lib/action_view/template/renderable.rb [hunk 2]
```diff
@@ -4,20 +4,22 @@ module ActionView
...
       end
 
       def format
-        @renderable.format
+        @renderable.try(:format) || Array(@formats).first
```
Comment (rafaelfranca) [inline actionview/lib/action_view/template/renderable.rb:22]: Why are we trying to call the format now?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: actionview/lib/action_view/template/renderable.rb (hunk 1 above)
Comment (joelhawksley) [reply actionview/lib/action_view/template/renderable.rb:18]: @rafaelfranca I could also see Rails checking the arity?

Regardless, it'd be a huge bummer to have a breaking change like this. I'd _really_ hope we could avoid going down that route.

Comment (rafaelfranca) [reply actionview/lib/action_view/template/renderable.rb:18]: Yeah, I was thinking on doing the arity check to show the deprecation.

Why would it be a bummer if we deprecated the old signature? The old code would work, showing a deprecation, and if the library changed to the new one, as the arguments are optional, things would just work in old versions of Rails. In fact you can release a new version of a library right now with the new signature.

File: actionview/lib/action_view/template/renderable.rb (hunk 2 above)
Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb:22]: @rafaelfranca this is related to https://github.com/rails/rails/pull/50622.

When `#render_in` was introduced, there wasn't any public documentation citing that in addition to `#render_in`, objects would also need to define `#format`.

The `TestRenderable` in the test suite defines it (and, for example, I think `ViewComponent::Base` also defines it). It's called internal to Action View, so it's part of the public API despite the fact that it wasn't part of any of the published documentation.

https://github.com/rails/rails/pull/50622/files#diff-70344644d049724d12c9a0c2c387dd1ea31663fbfad4c6a1acdd49efc3cb9512R20 is actually the source of the `Object#try` change. The motivation behind that change was that since it was newly documented, prior implementers of `#render_in` could get by without defining `#format`. The absence of the method means that `TemplateRenderable#format` returns `nil`, and somewhere down the chain the content gets treated as `:html` by default.

Since this change makes more of the options available to the `#render_in` call, it could be useful for the caller to decide how to render itself (as HTML, JSON, etc) based on that newly available information.

If the `Object#try` doesn't feel correct, would we be able to change this interface to accept the `:formats` available from the `render` call as arguments?

File: actionview/lib/action_view/template/renderable.rb [hunk 3]
```diff
@@ -4,20 +4,31 @@ module ActionView
...
         @renderable.class.name
       end
 
-      def render(context, *args)
-        @renderable.render_in(context)
+      def render(context, locals)
+        case @renderable.method(:render_in).arity
+        when 1
+          ActionView.deprecator.warn <<~WARN
+            #render_in signatures without options are deprecated.
```
Comment (seanpdoyle) [inline actionview/lib/action_view/template/renderable.rb:21]: @rafaelfranca @joelhawksley I've added a deprecation here along with test coverage. I struggled to convey a more detailed deprecation message. If you're able, please share any suggestions you have.

File: guides/source/layouts_and_rendering.md
Comment (seanpdoyle) [inline guides/source/layouts_and_rendering.md:1]: I think these examples could be improved, but I'm not entirely sure how.

In my mind, there's a few points worth highlighting:

1. `#render_in` and the `:renderable` key as an alternative to `:partial` is novel in its own right.
2. The object that defines `#render_in` can decide on its own how to mix and match its own internal state with the state provided at the `render` call site through the `:locals` options
3. The object that defines `#render_in` can also decide how to mix the block passed to `render`
4. The object can decide which format to render into

Prior to this commit, 1. is only point mentioned at all.

With the proposed changes, 2. poses an interesting opportunity. I consider it similar to React's split between State and Props. Maybe there's an `Avatar` object that manages its own internal state like the URL and alt text, but accepts overrides from `:locals` like `:size`:

```ruby
class Avatar
  include ActiveModel::Model
  include ActiveModel::Attributes

  attribute :url, :string
  attribute :alt, :string
  attribute :size, :string, default: "medium"

  def render_in(view_context, **options)
    # ...
  end
end

avatar = Avatar.new(url: "https://example.com/avatar.jpg", alt: "A smiling person")

render avatar
# => <img src="https://example.com/avatar.jpg" alt="A smiling person" class="avatar avatar--medium">
render avatar, size: "small"
# => <img src="https://example.com/avatar.jpg" alt="A smiling person" class="avatar avatar--small">
```

Block-centric rendering objects like `ViewComponent::Base` could benefit from 3. by forwarding the `render` call's block to its own constructor. Prior to this commit, calls like this wouldn't behave like you'd expect (as highlighted by https://github.com/rails/rails/pull/45432):

```ruby
# ignores the block
render MyComponent.new do
  "Content"
end

# ignores the block
render(MyComponent.new) do
  "Content"
end

# captures the block
render(MyComponent.new do
  "Content"
end)
```

Support for determining the templating format (4.) is potentially desirable for objects that want to support multiple formats (like HTML, JSON, etc.). Unfortunately, that support might require more broad changes, and need to be incorporated in a subsequent PR. At the moment, it isn't clear the precedent between: 

* `view_context.lookup_context.formats`
* the `:formats` option passed to `render`
* the `#format` method defined by the same object as `#render_in`. At this point in time, the `#format` call would lack access to both the `view_context` and the render `options`

There's even potential for applications (or Rails itself) to extend models to know how to render themselves, if they want more control than what Action View provides out of the box:

```ruby
class ApplicationRecord < ActiveRecord::Base
  # Renders the object into an Action View context.
  #
  #   # app/models/person.rb
  #   class Person < ApplicationRecord
  #   end
  #
  #   # app/views/people/_person.html.erb
  #   <p><%= person.name %></p>
  #
  #   person = Person.new name: "Ralph"
  #
  #   render(person)              # => "<p>Ralph</p>
  #   render(renderable: person)  # => "<p>Ralph</p>
  def render_in(view_context, **options, &block)
    view_context.render(partial: to_partial_path, object: self, **options, &block)
  end
end
```

File: actionview/lib/action_view/template/renderable.rb (hunk 2 above)
Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb]: @rafaelfranca based on [this discussion](https://github.com/rails/rails/pull/50622#discussion_r1445897182), I've made two changes:

* omit support for `render formats: ...` from this PR
* deprecation warning for `@renderable` objects that don't respond to `#format`

Comment (rafaelfranca) [reply actionview/lib/action_view/template/renderable.rb]: > deprecation warning for @renderable objects that don't respond to #format
Why we need that?

File: actionpack/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,25 @@
+*   Accept render options and block in `render` calls made with `:renderable`
+
+    ```ruby
+    class Greeting
```
Comment (rafaelfranca) [inline actionpack/CHANGELOG.md:4]: Don't this object needs to respond to `format`?

File: actionview/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,26 @@
+*   Pass render options and block to calls to `#render_in`
+
+    ```ruby
+    class Greeting
```
Comment (rafaelfranca) [inline actionview/CHANGELOG.md:4]: Don't this object needs to respond to `format`?

File: actionview/lib/action_view/template/renderable.rb [hunk 6]
```diff
@@ -4,22 +4,44 @@ module ActionView
...
         raise ArgumentError, "'#{@renderable.inspect}' is not a renderable object. It must implement #render_in."
       end
 
       def format
-        @renderable.try(:format)
+        if @renderable.respond_to?(:format)
+          @renderable.format
+        else
+          ActionView.deprecator.warn <<~WARN
+            #{@renderable.inspect} must respond to `#format` to be renderable
```
Comment (rafaelfranca) [inline actionview/lib/action_view/template/renderable.rb:40]: Documentation in a bunch of places on this PR says `format` is optional, here we are telling it is required. Which is the right requirement? If we need format, why?

Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb:40]: I agree with you, the documentation should be specific about the fact that `#format` must be implemented. I'll make those changes now.

I've added this deprecation because the requirement for `#format` was never mentioned prior to https://github.com/rails/rails/pull/50622/files.

Is requiring that `#format` be implemented considered a breaking change, since there might be apps that have `#render_in` objects that don't define `#format`?

Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb:40]: On the other hand, the code _does not_ current require that `#format` be implemented thanks to https://github.com/rails/rails/pull/50622/files#r1445897182.

We could continue to treat `#format` as optional, in which case the deprecation would not be necessary, and the code could remain `@renderable.try(:format)`.

Does that seem preferred over a deprecation?

Comment (rafaelfranca) [reply actionview/lib/action_view/template/renderable.rb:40]: Before #50622, `format` was already required. That PR made it optional, so there is no change in behavior there that needs to be deprecated.

Now, if #50622 changed from `@renderable.try(:format)` to `@renderable.format` we would need a deprecation. The opposite direction doesn't need a deprecation.

Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb:40]: @rafaelfranca thank you for clarifying. 

Since #50622 made that optional, I will change the documentation to reinforce that it is optional, then remove the deprecation.

Comment (rafaelfranca) [reply actionview/lib/action_view/template/renderable.rb:40]: 👍🏽

File: actionview/test/template/render_test.rb [hunk 7]
```diff
@@ -312,11 +312,11 @@ def test_render_renderable_with_incompatible_object
...
   end
 
-  def test_render_renderable_does_not_mask_nomethoderror_from_within_render_in
+  def test_render_renderable_does_not_mask_nameerror_from_within_render_in
     renderable = Object.new
-    renderable.define_singleton_method(:render_in) { |*| nil.foo }
+    renderable.define_singleton_method(:render_in) { |*| nil.method(:foo) }
 
-    assert_raises NoMethodError, match: "undefined method `foo' for nil" do
+    assert_raises NameError, match: "undefined method `foo' for class `NilClass'" do
```
Comment (jonathanhefner) [inline actionview/test/template/render_test.rb:319]: I don't think this should change.  Presumably, we will remove the arity check when we remove the deprecated code, and go back to rescuing `NoMethodError`.  If we forget to also change this test back, we will lose test coverage.

Conversely, raising `NoMethodError` in this test can exercise both `rescue NameError` and `rescue NoMethodError`.

**Comment by BlakeWilliams** [comment]
> I can relate to that! Since ViewComponent was the pilot tool for driving out `render_in`, that makes a lot of sense. In my understanding of where ViewComponent fits in, its constructor state **is** its render context. That's also reasonable, since it's aiming to serve as a View Model. In my mind, ViewComponent has an opportunity to take a stance on requiring constructor-time state and rejecting render-time locals.

I wouldn't want us to take that stance since that creates a pretty awkward API for consumers since there's two supported paths but only one is correct. I chatted with the rest of the ViewComponent team and I think we're in agreement that this API change is fine, _but_ it will require us to make changes in ViewComponent to properly support it (which I'm somewhat excited about).

If this change is approved we'll need to update the library to support both paths:

```ruby
render(MyComponent, name: "Fox Mulder")
render(MyComponent.new(name: "Fox Mulder"))
```

I'm excited because the first example above could be used to enable component authors to omit an `initializer`, which is something I've been wanting to see for a while but couldn't quite fit it in the existing API in a reasonable way. The second example would likely take that stance of dropping locals if we continue supporting it (which we likely will but I'm tempted to propose we don't).

So tl;dr is that this change seems fine (if not good) from a ViewComponent perspective. 👍

File: actionview/lib/action_view/template/renderable.rb [hunk 8]
```diff
@@ -4,17 +4,31 @@ module ActionView
...
+              Change #render_in to accept keyword arguments.
+            WARN
+
+            {}
+          else
+            { locals: locals }
+          end
+
+        @renderable.render_in(context, **options, &@block)
+      rescue NameError
```
Comment (seanpdoyle) [inline actionview/lib/action_view/template/renderable.rb:31]: @jonathanhefner since you helped me workshop the `NameError` handling below specific to `NameError` instances that stem from `#render_in` and **only** `#render_in`, I'm curious if a similar approach would be safe for `ArgumentError` handling:

```suggestion
      def render(context, locals)
        @renderable.render_in(context, locals: locals, &@block)
      rescue ArgumentError
        if @renderable.method(:render_in).arity == 1
          ActionView.deprecator.warn <<~WARN
            Action View support for #render_in without options is deprecated.

            Change #render_in to accept keyword arguments.
          WARN

          @renderable.render_in(context, &@block)
        else
          raise
        end
      rescue NameError
```

The motivation here is to optimize adherents to the interface to avoid slowing the happy path down with conditionals. Does that feel like a good trade-off?

File: actionview/lib/action_view/template/renderable.rb (hunk 1 above)
Comment (seanpdoyle) [reply actionview/lib/action_view/template/renderable.rb]: If this is a breaking change, would it need to be part of a major version like 7.2 to 8.0?

If not, is the deprecation in the current implementation sufficient enough for a future minor release (like 8.0 to 8.1)?

**Review by rafaelfranca — APPROVED** [review]

**Comment by seanpdoyle** [comment]
Thank you for the final review @rafaelfranca !

@joelhawksley @BlakeWilliams I've opened https://github.com/rails/rails/pull/57349 as a follow-up to this change to provide an opportunity to bridge from Active Model to other view layer integrations.

**Comment by flavorjones** [comment]
@seanpdoyle Just a note that libraries like Lexxy that call `super` in their `render_in` method like this:

```ruby
  class Editor::LexxyEditor::Tag < Editor::Tag
    def render_in(view_context, ...)
      options[:value] = options[:value].to_str if options[:value].respond_to?(:to_str)
      super
    end
  end
```

will get an exception when `super` invokes `Editor::Tag#render_in`:

```
ArgumentError: wrong number of arguments (given 2, expected 1)
```

I've opened https://github.com/rails/rails/pull/57356 to fix it.

File: actionview/lib/action_view/template/renderable.rb [hunk 9]
```diff
@@ -4,17 +4,28 @@ module ActionView
...
 
       def identifier
         @renderable.class.name
       end
 
-      def render(context, *args)
-        @renderable.render_in(context)
-      rescue NoMethodError
+      def render(context, locals)
+        if @renderable.method(:render_in).arity == 1
```
Comment (chaadow) [inline actionview/lib/action_view/template/renderable.rb:17]: @seanpdoyle  @rafaelfranca Just FYI, i had an issue with a renderable having an `attr_reader :method` ( i'm using it as the HTTP method for my links) 

and so was having `wrong number of arguments (given 1, expected 0)`

renaming my attr_reader from `method` to `http_method` fixed the issue, but thought I would tell you about it in case maybe you want to switch to an UnboundedMethod then bind the `@renderable`

Your call! 🫡

Comment (rafaelfranca) [reply actionview/lib/action_view/template/renderable.rb:17]: Can you open a pr? I don’t want people to have to change the code

Comment (chaadow) [reply actionview/lib/action_view/template/renderable.rb:17]: @rafaelfranca Yes I will open a PR later, thanks for confirming.

Comment (chaadow) [reply actionview/lib/action_view/template/renderable.rb:17]: Here is the PR: https://github.com/rails/rails/pull/57360 @rafaelfranca
